### PR TITLE
Stop infinite retries of upload and clean up snapshot

### PR DESCRIPTION
**Pull request by lipingxue**
**What this PR does / why we need it**:
Currently, we Infinite retry of upload in the Data Manager.  When upload keeps failing , the local snapshot was not deleted.  We need to delete snapshots if upload keeps failing.
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```
With this change, users can configure the "upload-cr-retry-max" in  configmap "velero-vsphere-plugin-config".

If this parameter is not specified, the default value 10 will be used.  Upload will retry until it reaches the 
maximum retry. 

After that, we will try to delete the local snapshot and mark the upload cr as "UploadFailedAfterRetry" and 
stop further retry.
```

### CODE REVIEW

**Comment by lipingxue** [comment]
Testing is done. See this 
[pr-test-result.txt](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/files/11510012/pr-test-result.txt)

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/806/

File: pkg/apis/backupdriver/v1alpha1/snapshot.go [hunk 1]
```diff
@@ -32,16 +32,17 @@ type SnapshotSpec struct {
...
+	SnapshotPhaseInProgress             SnapshotPhase = "InProgress"
+	SnapshotPhaseSnapshotted            SnapshotPhase = "Snapshotted"
+	SnapshotPhaseSnapshotFailed         SnapshotPhase = "SnapshotFailed"
+	SnapshotPhaseUploading              SnapshotPhase = "Uploading"
+	SnapshotPhaseUploaded               SnapshotPhase = "Uploaded"
+	SnapshotPhaseUploadFailed           SnapshotPhase = "UploadFailed"
+	SnapshotPhaseCanceling              SnapshotPhase = "Canceling"
+	SnapshotPhaseCanceled               SnapshotPhase = "Canceled"
+	SnapshotPhaseCleanupFailed          SnapshotPhase = "CleanupAfterUploadFailed"
+	SnapshotPhaseUploadFailedAfterRetry SnapshotPhase = "UploadFailedAfterRetry"
```
Comment (xing-yang) [inline pkg/apis/backupdriver/v1alpha1/snapshot.go:45]: Will be good to add comments to explain how SnapshotPhaseUploadFailedAfterRetry is different from SnapshotPhaseUploadFailed.  Why do we need both?

File: pkg/apis/datamover/v1alpha1/upload.go [hunk 2]
```diff
@@ -44,17 +44,18 @@ type UploadSpec struct {
...
-	UploadPhaseCanceling     UploadPhase = "Canceling"
-	UploadPhaseCanceled      UploadPhase = "Canceled"
+	UploadPhaseNew                    UploadPhase = "New"
+	UploadPhaseInProgress             UploadPhase = "InProgress"
+	UploadPhaseCompleted              UploadPhase = "Completed"
+	UploadPhaseUploadError            UploadPhase = "UploadError"
+	UploadPhaseCleanupFailed          UploadPhase = "CleanupFailed"
+	UploadPhaseCanceling              UploadPhase = "Canceling"
+	UploadPhaseCanceled               UploadPhase = "Canceled"
+	UploadPhaseUploadFailedAfterRetry UploadPhase = "UploadFailedAfterRetry"
```
Comment (xing-yang) [inline pkg/apis/datamover/v1alpha1/upload.go:58]: Same comment as above.

File: pkg/constants/constants.go [hunk 3]
```diff
@@ -176,8 +176,9 @@ const (
 )
 
 const (
-	RetryInterval = 5
-	RetryMaximum  = 5
+	RetryInterval               = 5
+	RetryMaximum                = 5
+	DefaultUploadCRRetryMaximum = 10
```
Comment (xing-yang) [inline pkg/constants/constants.go:181]: Add comments to explain what happens when the max is reached.

File: pkg/constants/constants.go [hunk 4]
```diff
@@ -418,12 +419,13 @@ const (
...
 )
 
 const (
 	// AnnVolumeHealth is the key for HealthStatus annotation on volume claim
 	// for vSphere CSI Driver.
 	AnnVolumeHealth = "volumehealth.storage.kubernetes.io/health"
-        // key for expressing timestamp for volume health annotation
-        AnnVolumeHealthTS = "volumehealth.storage.kubernetes.io/health-timestamp"
+	// key for expressing timestamp for volume health annotation
+	AnnVolumeHealthTS = "volumehealth.storage.kubernetes.io/health-timestamp"
```
Comment (xing-yang) [inline pkg/constants/constants.go:430]: @deepakkinni Any idea why we need these in our constants?  These are vSphere CSI driver annotations.

File: pkg/controller/upload_controller.go [hunk 5]
```diff
@@ -306,6 +307,11 @@ func (c *uploadController) processUpload(req *pluginv1api.Upload) error {
 		return nil
 	}
 
+	if req.Status.Phase == pluginv1api.UploadPhaseUploadFailedAfterRetry {
+		log.WithField("phase", req.Status.Phase).WithField("generation", req.Generation).Debug("The status of upload CR in kubernetes API server is failed after retry. Skipping it")
```
Comment (xing-yang) [inline pkg/controller/upload_controller.go:311]: upload -> Upload
kubernetes -> Kubernetes

File: pkg/controller/upload_controller.go [hunk 6]
```diff
@@ -372,12 +378,37 @@ func (c *uploadController) processUpload(req *pluginv1api.Upload) error {
...
+			log.Infof("UploadCRRetryMaximum is %d", uploadCRRetryMaximum)
+			if req.Status.RetryCount < int32(uploadCRRetryMaximum) {
+				_, err = c.patchUploadByStatusWithRetry(req, pluginv1api.UploadPhaseUploadError, errMsg)
+				if err != nil {
+					errMsg = fmt.Sprintf("%v. %v", errMsg, errors.WithStack(err))
+				}
+				log.Error(errMsg)
+				return errors.New(errMsg)
+			} else {
+				log.Infof("Upload CR %s failed to upload snapshot after retry %d times", req.Name, uploadCRRetryMaximum)
```
Comment (xing-yang) [inline pkg/controller/upload_controller.go:391]: retry -> retrying

File: pkg/controller/upload_controller.go [hunk 7]
```diff
@@ -372,12 +378,37 @@ func (c *uploadController) processUpload(req *pluginv1api.Upload) error {
...
+				err = c.snapMgr.DeleteLocalSnapshot(peID)
+				if err != nil {
+					errMsg := fmt.Sprintf("Failed to clean up local snapshot, %s, error: %v. will retry.", peID.String(), errors.WithStack(err))
+					_, err = c.patchUploadByStatusWithRetry(req, pluginv1api.UploadPhaseCleanupFailed, errMsg)
+					if err != nil {
+						errMsg = fmt.Sprintf("Failed to patch upload status with retry, errror: %v", errors.WithStack(err))
```
Comment (xing-yang) [inline pkg/controller/upload_controller.go:397]: Upload

**Review by xing-yang — COMMENTED** [review]
Please also update the README accordingly.

File: pkg/apis/backupdriver/v1alpha1/snapshot.go (hunk 1 above)
Comment (lipingxue) [reply pkg/apis/backupdriver/v1alpha1/snapshot.go:45]: My understanding is each upload status  need to have corresponding snapshot status and we set the snapshot status based on upload status in function "syncUploadByKey".

Comment (lipingxue) [reply pkg/apis/backupdriver/v1alpha1/snapshot.go:45]: Please review the comments I added in the code.  After we agree the wording are fine, I will add them to README.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/808/

File: pkg/constants/constants.go (hunk 4 above)
Comment (deepakkinni) [reply pkg/constants/constants.go:437]: https://github.com/vmware-tanzu/velero-plugin-for-vsphere/blob/0f02ba2461e03f09773606c3cf316d509d3a6941/pkg/plugin/backup_pvc_action_plugin.go#L207-L211

File: pkg/utils/utils.go [hunk 8]
```diff
@@ -733,6 +733,49 @@ func GetSecretNamespaceAndName(kubeClient kubernetes.Interface, veleroNs string,
...
+	if config == nil {
+		config, err = GetKubeClientConfig()
+		if err != nil {
+			return constants.DefaultUploadCRRetryMaximum
+		}
+	}
+
+	clientset, err := GetKubeClientSet(config)
+	if err != nil {
+		return constants.DefaultUploadCRRetryMaximum
```
Comment (deepakkinni) [inline pkg/utils/utils.go:765]: log the error, someone debugging should be aware that the retry maximum was not honored.

File: pkg/utils/utils.go [hunk 9]
```diff
@@ -733,6 +733,49 @@ func GetSecretNamespaceAndName(kubeClient kubernetes.Interface, veleroNs string,
...
+	configMap := veleroPuginConfigMap.Data
+	if maxRetryCountStr, ok := configMap[constants.UploadCRRetryMaximumKey]; ok {
+		maxRetryCount, err := strconv.Atoi(maxRetryCountStr)
+		if err != nil {
+			fmt.Printf("UploadCRRretryMaximum %s is invalid, err: %+v.\n", maxRetryCountStr, err)
+			maxRetryCount = constants.DefaultUploadCRRetryMaximum
+		}
+		return maxRetryCount
+	}
+	return constants.DefaultUploadCRRetryMaximum
```
Comment (deepakkinni) [inline pkg/utils/utils.go:751]: Same, log it here as well.

File: pkg/controller/upload_controller.go [hunk 10]
```diff
@@ -372,12 +378,37 @@ func (c *uploadController) processUpload(req *pluginv1api.Upload) error {
...
+				} else {
+					msg := fmt.Sprintf("Successfully deleted local snapshot, %v", peID.String())
+					log.Info(msg)
+					// UploadPhaseUploadFailedAfterRetry is a terminal state, and upload CR will not retry
+					_, err = c.patchUploadByStatusWithRetry(req, pluginv1api.UploadPhaseUploadFailedAfterRetry, msg)
+					if err != nil {
+						log.Error(err)
+					}
+					return nil
+				}
```
Comment (deepakkinni) [inline pkg/controller/upload_controller.go:410]: This need not be in else block.

File: pkg/controller/upload_controller.go [hunk 11]
```diff
@@ -372,12 +378,37 @@ func (c *uploadController) processUpload(req *pluginv1api.Upload) error {
...
+			uploadCRRetryMaximum := utils.GetUploadCRRetryMaximum(nil)
+			log.Infof("UploadCRRetryMaximum is %d", uploadCRRetryMaximum)
+			if req.Status.RetryCount < int32(uploadCRRetryMaximum) {
+				_, err = c.patchUploadByStatusWithRetry(req, pluginv1api.UploadPhaseUploadError, errMsg)
+				if err != nil {
+					errMsg = fmt.Sprintf("%v. %v", errMsg, errors.WithStack(err))
+				}
+				log.Error(errMsg)
+				return errors.New(errMsg)
+			} else {
```
Comment (deepakkinni) [inline pkg/controller/upload_controller.go:390]: Try avoiding multiple if-else blocks. There are return statements in if block already.

**Review by deepakkinni — COMMENTED** [review]
Can you add some unit tests as well.

File: pkg/constants/constants.go [hunk 12]
```diff
@@ -178,6 +178,14 @@ const (
 const (
 	RetryInterval = 5
 	RetryMaximum  = 5
+	// Default maximum retry count for upload CR.
+	// After upload CR retry reaches the maximum and still cannot upload the snapshot to durable storage. Upload CR will stop further retry and
+	// try to delete the local snapshot.
+	// If local snapshot is deleted successfully, the status of upload CR will change to "UploadFailedAfterRetry".
+	// If local snapshot cannot be deleted, the stauts of upload CR will change to "CleanupFailed".
```
[the remark this answers is not in the document]
Comment (lipingxue) [reply pkg/constants/constants.go:185]: Update the comment of those states. Will update the .md file once this change is approved.

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/811/

**Comment by lipingxue** [comment]
> Can you add some unit tests as well.

Yes. I added two unit tests. This is the unit test result. 
[unit-test-May-23.txt](https://github.com/vmware-tanzu/velero-plugin-for-vsphere/files/11549196/unit-test-May-23.txt)

File: pkg/constants/constants.go (hunk 12 above)
Comment (xing-yang) [inline pkg/constants/constants.go:185]: typo: stauts

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/817/

**Review by deepakkinni — APPROVED** [review]

**Comment by lipingxue** [comment]
@xing-yang I have squashed the commits, please help to merge it. Thanks!

**Comment by svcbot-qecnsdp** [comment]
Sorry, some of the pre-check tests are failing for this PR ! 

Please check the Jenkins job for more details
https://container-dp.svc.eng.vmware.com/job/CNSDP-CI/818/

**Pull request merged**